### ISSUE TITLE
Typed int parameters with leading zeroes break

### ISSUE THREAD

**Issue by chriscpty** [yiisoft/yii2#20278]
### What steps will reproduce the problem?

Write a controller action like this (or similar):

```php
public function actionTest(int $test) {
    return $this->asJson(['res' => $test]);
}
```

Then call it with `?test=1` and `?test=01`.

### What is the expected result?

Calling this action with any number-shaped parameter should work, even with leading zeroes - `/{controller}/test?test=1` and `/{controller}/test?test=01` should yield the same result.

### What do you get instead?

Calling this route as `/{controller}/test?test=1` works fine, but calling it as  `/{controller}/test?test=01` throws a bad request error, stating that `test` contains invalid data.

I assume this is because `yii\web\Controller::bindActionParams` just calls `filter_var` to check if a variable is a valid integer:

```php
// yii\web\Controller:152
case 'int':
    $params[$name] = filter_var($params[$name], FILTER_VALIDATE_INT, FILTER_NULL_ON_FAILURE);
    break;
```

`FILTER_VALIDATE_INT` doesn't seem to accept leading zeroes in its input - this will probably need to be replaced with a custom filter function to fix this issue.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | dev-master
| PHP version      | 8.1 
| Operating system | Debian 12

**Comment by schmunk42** [comment]
In PHP `$foo = 011` is actually an octal notation, which has an integer value of `9`.

There's a flag for https://www.php.net/manual/en/function.filter-var.php called `FILTER_FLAG_ALLOW_OCTAL` (which I can not find anywhere in the PHP docs), but the constants are there:

![image](https://github.com/user-attachments/assets/59aed457-467b-43ba-a8d4-f4ea7b4b07f4)

See also [example](https://3v4l.org/WSgXn) for different notations and PHP versions.

--- 

TLDR;

I think you need to sanitize your values with an URL rule.

**Comment by chriscpty** [comment]
re use cases, I ran into it because a widget I used created values with leading zeroes, and i directly used those values as query params in ajax requests. My workaround for that right now is to cast the values to integers.

TIL about PHP's octal notation - I guess if yii2 wants to support use cases with leading zeroes, a custom function along the lines of `if (!is_numeric($input) {return null;}; return (int)$input` is needed, replacing the `filter_var` call?
